File: pr016.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
2 ** 15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 2 ** 1000?
"""

INPUT_NUM = 1000


def mult2(nlst):
  """calclulate double of nlst """
  over = 0
  for i in range(len(nlst)):
    v = nlst[i] * 2 + over
    over, nlst[i] = divmod(v, 10)
  if over > 0:
    nlst.append(over)


def bvalue(num):
  """
  return '2 ** num' as the digit list
  eg) bvalue(15) = [3, 2, 7, 6, 8] = 32768 = 2 ** 15
  """
  val = [1]
  for i in range(num):
    mult2(val)
  val.reverse()
  return val


def main(num):
  """main function"""
  print sum(bvalue(num))


if __name__ == "__main__":
  main(INPUT_NUM)

File: pr016.hs
{-
2 ** 15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 2 ** 1000?
-}

module Main where

input_num :: Int
input_num = 1000

-- double value of digit list.
mult2 :: [Int] -> Int -> [Int]
mult2 [] over = if over == 0
                  then []
                  else [over]
mult2 (n:ns) over = m:mult2 ns d
                      where
                        v = 2 * n + over
                        m = v `mod` 10
                        d = v `div` 10

-- subroutine of bvalue
bvalue_sub :: Int -> [Int] -> [Int]
bvalue_sub 0 ns = ns
bvalue_sub n ns = bvalue_sub (n-1) (mult2 ns 0)

-- return 2 ** n as digit list
bvalue :: Int -> [Int]
bvalue n = bvalue_sub n [1]

main :: IO ()
main = print $ sum $ bvalue input_num